File: 1056-capacity-to-ship-packages-within-d-days/solution.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int n = weights.length;
        int sum = 0;
        int maxVal = Integer.MIN_VALUE;
        for(int i=0; i<n; i++) {
            maxVal = Math.max(maxVal, weights[i]);
        }
        for(int i=0; i<n; i++) {
            sum += weights[i];
        }
        int low = maxVal;
        int high = sum;
        int ans = 0;

        while(low <= high) {
            int mid = low+(high-low)/2;//ship capacity
            if(canLoad(weights, mid, days)) {
                ans = mid;
                high = mid-1;
            } else {
                low = mid+1;
            }
        }
        return ans;
    }

    public boolean canLoad(int[] weights, int minCap, int days) {
        int totalCap = 0;
        int day = 1;
        int n = weights.length;
        for(int pack=0; pack < n; pack++) {
            totalCap += weights[pack];
            // # too heavy, wait for the next day
            if(totalCap > minCap) {
                totalCap = weights[pack];
                day++;
                if(day > days) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: 1056-capacity-to-ship-packages-within-d-days/solution.swift
class Solution {
    func shipWithinDays(_ weights: [Int], _ days: Int) -> Int {
        var n = weights.count
        var low = weights.max()!
        var sum = 0

        for weight in weights {
            sum += weight
        }
        var high = sum
        var ans = 0

        while(low <= high) {
            var mid = (low+high)/2
            if(canLoad(mid)) {
                ans = mid
                high = mid - 1
            } else {
                low = mid + 1
            }
        }

        func canLoad(_ capacity: Int) -> Bool {
            var totalPackages = 0
            var day = 1
            for weight in weights {
                totalPackages += weight
                if totalPackages > capacity {
                    // keep adding
                    totalPackages = weight
                    day += 1
                } 
                if(day > days) {
                    return false
                }      
            }
            return day <= days
        }
        return ans
    }
}
